pkg/tfgen/generate_csharp.go

```diff
@@ -0,0 +1,1312 @@
+		rg.w.Writefmtln("")
+	}
+	rg.w.Writefmtln("")
+
+	// Emit the class constructor.
+	argsType := rg.res.argst.name
+
+	var argsDefault string
+	if len(rg.res.reqprops) == 0 {
+		// If the number of required input properties was zero, we can make the args object optional.
+		argsDefault = " = null"
+		argsType += "?"
+	}
+
+	optionsType := "CustomResourceOptions"
+	if rg.res.IsProvider() {
+		optionsType = "ResourceOptions"
+	}
+
+	// Write a comment prior to the constructor.
+	rg.w.Writefmtln("        /// <summary>")
+	rg.w.Writefmtln("        /// Create a %s resource with the given unique name, arguments, and options.", className)
+	rg.w.Writefmtln("        ///")
+	rg.w.Writefmtln("        /// <param name=\"name\">The unique name of the resource</param>")
+	rg.w.Writefmtln("        /// <param name=\"args\">The arguments used to populate this resource's properties</param>")
+	rg.w.Writefmtln("        /// <param name=\"options\">A bag of options that control this resource's behavior</param>")
+	rg.w.Writefmtln("        /// </summary>")
+
+	rg.w.Writefmtln("        public %s(string name, %s args%s, %s? options = null)",
+		className, argsType, argsDefault, optionsType)
+	rg.w.Writefmtln("            : base(\"%s\", name, args, MakeResourceOptions(options, \"\"))", rg.res.info.Tok)
+	rg.w.Writefmtln("        {")
+	rg.w.Writefmtln("        }")
+
+	// Write a private constructor for the use of `Get`.
+	if !rg.res.IsProvider() {
+		rg.w.Writefmtln("")
+		rg.w.Writefmtln("        private %s(string name, Input<string> id, %s? state = null, %s? options = null)",
+			className, rg.res.statet.name, optionsType)
+		rg.w.Writefmtln("            : base(\"%s\", name, state, MakeResourceOptions(options, id))", rg.res.info.Tok)
+		rg.w.Writefmtln("        {")
+		rg.w.Writefmtln("        }")
+	}
+
+	// Write the method that will calculate the resource options.
+	rg.w.Writefmtln("")
+	rg.w.Writefmtln("        private static %[1]s MakeResourceOptions(%[1]s? options, Input<string>? id)", optionsType)
+	rg.w.Writefmtln("        {")
+	rg.w.Writefmtln("            var defaultOptions = new %s", optionsType)
+	rg.w.Writefmtln("            {")
+	rg.w.Writefmtln("                Id = id,")
+	rg.w.Writefmt("                Version = Utilities.Version")
+
+	switch len(rg.res.info.Aliases) {
+	case 0:
+		rg.w.Writefmtln("")
+	case 1:
+		rg.w.Writefmtln(",")
+		rg.w.Writefmt("                Aliases = { ")
+		rg.generateAlias(rg.res.info.Aliases[0])
+		rg.w.Writefmtln(" }")
+	default:
+		rg.w.Writefmtln(",")
+		rg.w.Writefmtln("                Aliases =")
+		rg.w.Writefmtln("                {")
+		for i, alias := range rg.res.info.Aliases {
+			rg.w.Writefmt("                    ")
+			rg.generateAlias(alias)
+			if i < len(rg.res.info.Aliases)-1 {
+				rg.w.Writefmtln(",")
+			}
+		}
+		rg.w.Writefmtln("                }")
+	}
+
+	rg.w.Writefmtln("            };")
+	rg.w.Writefmtln("            return (%s)ResourceOptions.Merge(defaultOptions, options);", optionsType)
```

Comment:
ah, pity.  so i should make a `CustomResourceOptions.Merge` so you don't need to cast here.

Reply:
Note: you can now call `CustomResourceOptions.Merge`.  so you don't need the cast anymore.